src/QsCompiler/CompilerOptimizations/CallableInlining.fs

```diff
@@ -0,0 +1,183 @@
+    | Some (Provided (specArgs, body)), _
+    | Some (Generated SelfInverse), Some (Provided (specArgs, body)) -> Some (specArgs, body)
+    | _ -> None
+
+
+/// Stores all the data needed to inline a callable
+type private InliningInfo = {
+    functors: Functors
+    callable: QsCallable
+    arg: TypedExpression
+    specArgs: QsArgumentTuple
+    body: QsScope
+}
+
+/// Tries to construct an InliningInfo from the given expression.
+/// Returns None if the expression is not a callable invocation that can be inlining.
+let private tryGetInliningInfo callables expr =
+    maybe {
+        let! functors, callable, arg = trySplitCall callables expr
+        let! specArgs, body = tryGetProvidedImpl callable functors
+        return { functors = functors; callable = callable; arg = arg; specArgs = specArgs; body = body }
+    }
+
+
+/// Recursively finds all the callables that could be inlined into the given scope.
+/// Includes callables that could be indirectly called or inlined into this scope.
+/// Mutates the given HashSet by adding all the found callables to the set.
+/// Is used to prevent inlining recursive functions into themselves forever.
+let rec private findAllCalls (callables: Callables) (scope: QsScope) (found: HashSet<QsQualifiedName>): unit =
+    scope |> findAllBaseStatements |> Seq.iter (function
+        | QsExpressionStatement ex ->
+            match tryGetInliningInfo callables ex.Expression with
+            | Some ii ->
+                // Only recurse if we haven't processed this callable yet
+                if found.Add ii.callable.FullName then
+                    findAllCalls callables ii.body found
+            | None -> ()
+        | _ -> ())
+
+/// Returns whether the given callable could eventually inline the given callable.
+/// Is used to prevent inlining recursive functions into themselves forever.
+let private cannotReachCallable (callables: Callables) (scope: QsScope) (cannotReach: QsQualifiedName) =
+    let mySet = HashSet()
+    findAllCalls callables scope mySet
+    not (mySet.Contains cannotReach)
+
+    
+/// The SyntaxTreeTransformation used to inline callables
+type internal CallableInliner(callables) =
+    inherit OptimizingTransformation()
+
+    // The current callable we're in the process of transforming
+    let mutable currentCallable: QsCallable option = None
+    // The list of new statements to add
+    let mutable statementsToAdd = []
+
+    /// Inline an expression as a ScopeStatement, with many checks to ensure correctness.
+    /// Returns None if the expression cannot be safely inlined.
+    let safeInline expr =
+        maybe {
+            let! ii = tryGetInliningInfo callables expr
+            
+            do! check (countReturnStatements ii.body = 0)
+            let! current = currentCallable
+            do! check (cannotReachCallable callables ii.body current.FullName)
+            do! check (ii.functors.controlled < 2)
```

Comment:
What is the reason for this restriction?

Reply:
I check there is at most one Controlled functor because handling multiple Controlled functors means that I have to pull out all the controls from the nested argument tuple and combine them into a single array, which I haven't gotten around to implementing yet.

Reply:
Ok. How about a adding todo then?